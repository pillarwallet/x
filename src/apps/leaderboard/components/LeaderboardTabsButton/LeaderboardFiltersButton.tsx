type LeaderboardFiltersButtonProps = {
  isActive: boolean;
  text: string;
  onClick: () => void;
};

const LeaderboardFiltersButton = ({
  isActive,
  text,
  onClick,
}: LeaderboardFiltersButtonProps) => {
  return (
    <button
      type="button"
      className={`flex py-[7px] px-2 w-fit h-fit items-center justify-center ${isActive ? 'bg-dark_blue text-white' : 'bg-container_grey text-white/[.5]'} border-x-[1px] border-t-[1px] border-b-[3px] rounded-lg border-[#121116] cursor-pointer`}
      onClick={onClick}
    >
      <p className="font-normal text-[13px]">{text}</p>
    </button>
  );
};

export default LeaderboardFiltersButton;
