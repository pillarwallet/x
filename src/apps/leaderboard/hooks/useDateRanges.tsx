import { DateTime } from 'luxon';
import { useMemo } from 'react';

export const useDateRanges = () => {
  const now = DateTime.now();
  return useMemo(() => {
    return {
      currentMonday: now.startOf('week').toUnixInteger(),
      currentSunday: now.endOf('week').toUnixInteger(),
      lastMonday: now.startOf('week').minus({ weeks: 1 }).toUnixInteger(),
      lastSunday: now.endOf('week').minus({ weeks: 1 }).toUnixInteger(),
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [now.weekNumber]); // re-evaluate when the week changes
};
