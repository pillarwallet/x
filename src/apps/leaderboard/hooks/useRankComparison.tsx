import { useCallback } from 'react';

// types
import { LeaderboardRankChange } from '../../../types/api';

export const useRankComparison = () => {
  return useCallback((currentIndex: number, previousIndex: number) => {
    if (currentIndex === previousIndex) return LeaderboardRankChange.NO_CHANGE;
    return currentIndex > previousIndex
      ? LeaderboardRankChange.INCREASED
      : LeaderboardRankChange.DECREASED;
  }, []);
};
