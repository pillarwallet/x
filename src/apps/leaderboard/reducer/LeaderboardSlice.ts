/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export type LeaderboardTabsType = 'all' | 'migration' | 'trading';
export type LeaderboardTimeTabsType = 'all' | 'weekly';

export type LeaderboardState = {
  activeTab: LeaderboardTabsType;
  timeTab: LeaderboardTimeTabsType;
  isUserInMigrationData: boolean;
};

const initialState: LeaderboardState = {
  activeTab: 'all',
  timeTab: 'all',
  isUserInMigrationData: false,
};

const leaderboardSlice = createSlice({
  name: 'leaderboard',
  initialState,
  reducers: {
    setActiveTab(state, action: PayloadAction<LeaderboardTabsType>) {
      state.activeTab = action.payload;
    },
    setTimeTab(state, action: PayloadAction<LeaderboardTimeTabsType>) {
      state.timeTab = action.payload;
    },
    setIsUserInMigrationData(state, action: PayloadAction<boolean>) {
      state.isUserInMigrationData = action.payload;
    },
  },
});

export const { setActiveTab, setTimeTab, setIsUserInMigrationData } =
  leaderboardSlice.actions;

export default leaderboardSlice;
