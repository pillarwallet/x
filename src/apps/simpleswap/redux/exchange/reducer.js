import { createReducer } from '@reduxjs/toolkit';

import * as actions from './actions';

const initialState = {
  allCurrenciesList: {
    data: null,
    isLoading: true,
  },
  filter: {
    data: null,
    isLoading: true,
  },
  fixedCurrencies: null,
  allPairsList: null,
  fixedPairs: null,
  fiat: null,
  popularCurrencies: null,
  popularFiat: null,
  exchangesList: [],
};

const reducer = createReducer(initialState, (builder) => {
  builder.addCase(actions.setAllCurrencies, (state, action) => {
    const { data, isLoading } = action.payload;
    if (isLoading)
      return {
        ...state,
        allCurrenciesList: { data: null, isLoading: true },
      };

    const mapCoins = data.filter((coinItem) => !coinItem.isFiat);
    return {
      ...state,
      allCurrenciesList: { data: mapCoins, isLoading: false },
    };
  })
    .addCase(actions.setAllPairs, (state, action) => {
      const { data } = action.payload;

      return {
        ...state,
        allPairsList: data,
      };
    })
    .addCase(actions.setFixedCurrencies, (state, action) => {
      const { data } = action.payload;

      return {
        ...state,
        fixedCurrencies: data,
      };
    })
    .addCase(actions.setFixedPairs, (state, action) => {
      const { data: fixedPairs } = action.payload;

      return {
        ...state,
        fixedPairs,
      };
    })
    .addCase(actions.setPopularCurrencies, (state, action) => {
      const { popularCurrencies } = action.payload;

      return {
        ...state,
        popularCurrencies,
      };
    })
    .addCase(actions.setExchangeInfo, (state, action) => {
      const { data, exchangeId } = action.payload;

      const { exchangesList } = state;

      if (data.error) {
        return state;
      }

      const newExchangesList = exchangesList.filter((i) => i.id !== exchangeId);
      newExchangesList.push(data.data);

      return {
        ...state,
        exchangesList: newExchangesList,
      };
    })
})


export default reducer;
