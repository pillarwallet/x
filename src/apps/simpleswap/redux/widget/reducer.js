/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from '../../helpers/axios';
import {widgetInfo} from '../../constants/widgetInfo';

const initialState = {
  exchangeInfo: null,
};

export const GET_WIDGET_EXCHANGE_INFO = 'widget/GET_WIDGET_EXCHANGE_INFO';

// eslint-disable-next-line consistent-return
export const getWidgetExchangeInfo = createAsyncThunk(
  GET_WIDGET_EXCHANGE_INFO,
  async ({ exchangeId }) => {
      const headers = {
        'x-api-key': widgetInfo.apiKey,
      };
      const response = await axios.get(`/exchanges/${exchangeId}`, {
        headers,
      });

      return response.data.result;
  },
);

export const widgetSlice = createSlice({
  name: 'widget',
  initialState,
  reducers: {
    resetWidgetExchangeInfo: (state) => {
      state.exchangeInfo = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getWidgetExchangeInfo.fulfilled, (state, { payload }) => {
        state.exchangeInfo = payload;
      })
  },
});

export const { resetWidgetExchangeInfo } = widgetSlice.actions;

export default widgetSlice.reducer;
