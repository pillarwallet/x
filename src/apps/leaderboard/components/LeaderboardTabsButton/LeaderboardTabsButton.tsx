// reducer
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';
import {
  LeaderboardTimeTabsType,
  setTimeTab,
} from '../../reducer/LeaderboardSlice';

const LeaderboardTabsButton = () => {
  const dispatch = useAppDispatch();
  const timeTab = useAppSelector(
    (state) => state.leaderboard.timeTab as LeaderboardTimeTabsType
  );

  return (
    <div
      id="leaderboard-button-tabs"
      className="flex w-fit p-1 bg-[#121116] rounded-[10px]"
    >
      <button
        type="button"
        className={`items-center justify-center w-[130px] rounded-md p-1.5 ${timeTab === 'weekly' ? 'bg-container_grey' : 'bg-[#121116]'}`}
        onClick={() => dispatch(setTimeTab('weekly'))}
      >
        <p
          className={`font-normal text-[13px] ${timeTab === 'weekly' ? 'text-white' : 'text-white/[.5]'}`}
        >
          Weekly
        </p>
      </button>
      <button
        type="button"
        className={`items-center justify-center w-[130px] rounded-md p-1.5 ${timeTab === 'all' ? 'bg-container_grey' : 'bg-[#121116]'}`}
        onClick={() => dispatch(setTimeTab('all'))}
      >
        <p
          className={`font-normal text-[13px] ${timeTab === 'all' ? 'text-white' : 'text-white/[.5]'}`}
        >
          All Time
        </p>
      </button>
    </div>
  );
};

export default LeaderboardTabsButton;
