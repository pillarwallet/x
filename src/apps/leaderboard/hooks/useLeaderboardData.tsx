import { useEffect, useMemo, useState } from 'react';

// types
import {
  LeaderboardRankChange,
  LeaderboardTableData,
} from '../../../types/api';

// api queries
import {
  useGetLeaderboardMigrationQuery,
  useGetLeaderboardQuery,
} from '../api/leaderboard';

// utils
import {
  getCurrentWeekMigrationData,
  getLastWeekMigrationData,
  getMergeLeaderboardData,
  getMergeLeaderboardMigrationDataByAddresses,
} from '../utils';

// hooks
import { useDateRanges } from './useDateRanges';
import { useRankComparison } from './useRankComparison';

export const useLeaderboardData = () => {
  const dateRanges = useDateRanges();
  const compareIndexes = useRankComparison();
  const [weeklyTradingData, setWeeklyTradingData] = useState<
    LeaderboardTableData[] | undefined
  >(undefined);

  // API calls
  const allTimeQuery = useGetLeaderboardQuery(
    { skipDefaultPoints: true },
    { refetchOnMountOrArgChange: true, refetchOnFocus: true }
  );

  const weeklyQuery = useGetLeaderboardQuery(
    {
      skipDefaultPoints: true,
      from: dateRanges.currentMonday,
      until: dateRanges.currentSunday,
    },
    { refetchOnMountOrArgChange: true, refetchOnFocus: true }
  );

  const lastWeeklyQuery = useGetLeaderboardQuery(
    {
      skipDefaultPoints: true,
      from: dateRanges.lastMonday,
      until: dateRanges.lastSunday,
    },
    { refetchOnMountOrArgChange: true, refetchOnFocus: true }
  );

  const migrationQuery = useGetLeaderboardMigrationQuery(undefined, {
    refetchOnMountOrArgChange: true,
    refetchOnFocus: true,
  });

  // Process all time trading data
  const allTimeTradingData: LeaderboardTableData[] | undefined = useMemo(() => {
    if (!allTimeQuery.data?.results) return undefined;
    return allTimeQuery.data.results.map((result) => ({
      totalPoints: result.totalSwapAmountUsd || 0,
      totalAmountUsd: result.totalSwapAmountUsd || 0,
      totalGas: result.totalTxFeesUsd || 0,
      addresses: [result.address],
      newDropTime: result.pointsUpdatedAt || 0,
      finalPxPointsAwardEligible: result.finalPxPointsAwardEligible,
    }));
  }, [allTimeQuery.data]);

  // Process all time migration data
  const allTimeMigrationData: LeaderboardTableData[] | undefined =
    useMemo(() => {
      if (!migrationQuery.data?.result) return undefined;

      const allMigrationEntries = migrationQuery.data.result.map((result) => ({
        totalPoints: result?.progress?.totalPoints || 0,
        totalAmountUsd: result?.progress?.migratedAmountUsd || 0,
        addresses: result?.pxAddresses,
        totalGas: result?.progress?.migratedFeesUsd || 0,
        source: result.source || undefined,
      }));

      return getMergeLeaderboardMigrationDataByAddresses(allMigrationEntries);
    }, [migrationQuery.data]);

  // Process weekly trading data
  useEffect(() => {
    if (
      weeklyQuery.data &&
      lastWeeklyQuery.data &&
      !weeklyQuery.isLoading &&
      !lastWeeklyQuery.isLoading
    ) {
      const lastWeeklyDataMap = new Map<string, number>();
      lastWeeklyQuery.data.results.forEach((userData, index) => {
        lastWeeklyDataMap.set(userData.address, index);
      });

      const updatedWeeklyData: LeaderboardTableData[] | undefined =
        weeklyQuery.data.results.map((currentUserData, currentIndex) => {
          const lastIndex = lastWeeklyDataMap.get(currentUserData.address);
          const rankChange =
            lastIndex !== undefined
              ? compareIndexes(currentIndex, lastIndex)
              : LeaderboardRankChange.NO_CHANGE;

          return {
            totalPoints: currentUserData.totalSwapAmountUsd || 0,
            totalAmountUsd: currentUserData.totalSwapAmountUsd || 0,
            totalGas: currentUserData.totalTxFeesUsd || 0,
            addresses: [currentUserData.address],
            completedSwap: currentUserData.completedSwap || false,
            rankChange,
            finalPxPointsAwardEligible:
              currentUserData.finalPxPointsAwardEligible,
          };
        });

      setWeeklyTradingData(updatedWeeklyData);
    }
  }, [
    weeklyQuery.data,
    lastWeeklyQuery.data,
    weeklyQuery.isLoading,
    lastWeeklyQuery.isLoading,
    compareIndexes,
  ]);

  // Process weekly migration data
  const weeklyMigrationData: LeaderboardTableData[] | undefined =
    useMemo(() => {
      if (!migrationQuery.data?.result) return undefined;

      const currentEntries = getCurrentWeekMigrationData(
        migrationQuery.data.result
      );
      const lastWeekEntries = getLastWeekMigrationData(
        migrationQuery.data.result
      );

      if (!currentEntries || !lastWeekEntries) return undefined;

      const lastWeekMap = new Map<string, number>();
      lastWeekEntries.forEach((userData, index) => {
        const lastAddress = userData.addresses[userData.addresses.length - 1];
        if (lastAddress) lastWeekMap.set(lastAddress, index);
      });

      return currentEntries.map((currentUserData, currentIndex) => {
        const lastAddress =
          currentUserData.addresses[currentUserData.addresses.length - 1];
        const lastIndex = lastAddress
          ? lastWeekMap.get(lastAddress)
          : undefined;
        const rankChange =
          lastIndex !== undefined
            ? compareIndexes(currentIndex, lastIndex)
            : LeaderboardRankChange.NO_CHANGE;

        return { ...currentUserData, rankChange };
      });
    }, [migrationQuery.data, compareIndexes]);

  // Process all time merged data
  const mergedAllTimeData = useMemo(() => {
    if (!allTimeTradingData && !allTimeMigrationData) return undefined;
    return getMergeLeaderboardData(
      allTimeTradingData || [],
      allTimeMigrationData || []
    );
  }, [allTimeTradingData, allTimeMigrationData]);

  // Process weekly merged data
  const mergedWeeklyTimeData: LeaderboardTableData[] | undefined =
    useMemo(() => {
      if (
        !weeklyTradingData ||
        !weeklyMigrationData ||
        !lastWeeklyQuery.data?.results ||
        !migrationQuery.data?.result
      ) {
        return undefined;
      }

      // Create last week's merged data
      const lastWeekTradingData: LeaderboardTableData[] =
        lastWeeklyQuery.data.results.map((entry) => ({
          totalPoints: entry.totalSwapAmountUsd || 0,
          totalAmountUsd: entry.totalSwapAmountUsd || 0,
          totalGas: entry.totalTxFeesUsd || 0,
          addresses: [entry.address],
          completedSwap: entry.completedSwap === true,
        }));

      const lastWeekMigrationData = getLastWeekMigrationData(
        migrationQuery.data.result
      );

      const lastMerged = getMergeLeaderboardData(
        lastWeekTradingData,
        lastWeekMigrationData || []
      );

      // Create current week's merged data
      const currentMerged = getMergeLeaderboardData(
        weeklyTradingData,
        weeklyMigrationData
      );

      // Address to rank index map (based on last address)
      const lastIndexMap = new Map<string, number>();
      lastMerged.forEach((entry, index) => {
        // In merged data only one address (common to both trading and migration)
        const lastAddress = entry.addresses[0];
        if (lastAddress) lastIndexMap.set(lastAddress, index);
      });

      // Apply rank change
      const mergedWithRankChange = currentMerged.map((entry, currentIndex) => {
        // In merged data only one address (common to both trading and migration)
        const address = entry.addresses[0];
        const lastIndex = address ? lastIndexMap.get(address) : undefined;

        const rankChange =
          lastIndex !== undefined
            ? compareIndexes(currentIndex, lastIndex)
            : LeaderboardRankChange.NO_CHANGE;

        return {
          ...entry,
          rankChange,
        };
      });

      return mergedWithRankChange;
    }, [
      weeklyTradingData,
      weeklyMigrationData,
      lastWeeklyQuery.data,
      migrationQuery.data,
      compareIndexes,
    ]);

  return {
    // API Queries
    allTimeQuery,
    weeklyQuery,
    lastWeeklyQuery,
    migrationQuery,
    // Processed data
    allTimeTradingData,
    allTimeMigrationData,
    weeklyTradingData,
    weeklyMigrationData,
    mergedAllTimeData,
    mergedWeeklyTimeData,
    // Loading states
    isLoading: {
      allTime: allTimeQuery.isLoading,
      weekly: weeklyQuery.isLoading || lastWeeklyQuery.isLoading,
      migration: migrationQuery.isLoading,
      merged: allTimeQuery.isLoading || migrationQuery.isLoading,
      mergedWeekly:
        weeklyQuery.isLoading ||
        lastWeeklyQuery.isLoading ||
        migrationQuery.isLoading,
    },
    // Error states
    isError: {
      allTime: allTimeQuery.isError,
      weekly: weeklyQuery.isError || lastWeeklyQuery.isError,
      migration: migrationQuery.isError,
      merged: allTimeQuery.isError || migrationQuery.isError,
      mergedWeekly:
        weeklyQuery.isError ||
        lastWeeklyQuery.isError ||
        migrationQuery.isError,
    },
    // Success states
    isSuccess: {
      allTime: allTimeQuery.isSuccess,
      weekly: weeklyQuery.isSuccess && lastWeeklyQuery.isSuccess,
      migration: migrationQuery.isSuccess,
      merged: allTimeQuery.isSuccess && migrationQuery.isSuccess,
      mergedWeekly:
        weeklyQuery.isSuccess &&
        lastWeeklyQuery.isSuccess &&
        migrationQuery.isSuccess,
    },
  };
};
