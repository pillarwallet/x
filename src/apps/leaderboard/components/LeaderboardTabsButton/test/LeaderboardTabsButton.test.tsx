import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { vi } from 'vitest';

// store
import { store } from '../../../../../store';

// hooks
import * as reducerHooks from '../../../hooks/useReducerHooks';

// reducer
import { setTimeTab } from '../../../reducer/LeaderboardSlice';

// components
import LeaderboardTabsButton from '../LeaderboardTabsButton';

describe('<LeaderboardTabsButton />', () => {
  const useAppSelectorMock = vi.spyOn(reducerHooks, 'useAppSelector');
  const useAppDispatchMock = vi.spyOn(reducerHooks, 'useAppDispatch');

  const mockDispatch = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
    useAppDispatchMock.mockReturnValue(mockDispatch);
  });

  it('matches snapshot', () => {
    useAppSelectorMock.mockReturnValue('weekly');

    const { container } = render(
      <Provider store={store}>
        <LeaderboardTabsButton />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders both buttons', () => {
    useAppSelectorMock.mockReturnValue('weekly');

    render(
      <Provider store={store}>
        <LeaderboardTabsButton />
      </Provider>
    );

    expect(screen.getByRole('button', { name: /weekly/i })).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: /all time/i })
    ).toBeInTheDocument();
  });

  it('dispatches setTimeTab("all") when "All Time" is clicked', () => {
    useAppSelectorMock.mockReturnValue('weekly');

    render(
      <Provider store={store}>
        <LeaderboardTabsButton />
      </Provider>
    );

    fireEvent.click(screen.getByRole('button', { name: /all time/i }));
    expect(mockDispatch).toHaveBeenCalledWith(setTimeTab('all'));
  });

  it('dispatches setTimeTab("weekly") when "Weekly" is clicked', () => {
    useAppSelectorMock.mockReturnValue('all');

    render(
      <Provider store={store}>
        <LeaderboardTabsButton />
      </Provider>
    );

    fireEvent.click(screen.getByRole('button', { name: /weekly/i }));
    expect(mockDispatch).toHaveBeenCalledWith(setTimeTab('weekly'));
  });

  it('does not dispatch if clicking already active tab', () => {
    useAppSelectorMock.mockReturnValue('weekly');

    render(
      <Provider store={store}>
        <LeaderboardTabsButton />
      </Provider>
    );

    fireEvent.click(screen.getByRole('button', { name: /weekly/i }));

    expect(mockDispatch).toHaveBeenCalledWith(setTimeTab('weekly'));
  });
});
