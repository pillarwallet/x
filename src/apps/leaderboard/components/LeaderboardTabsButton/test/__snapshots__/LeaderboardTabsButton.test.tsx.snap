// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LeaderboardTabsButton /> renders correctly and matches snapshot 1`] = `
<div
  className="flex w-full bg-container_grey rounded-2xl p-4 gap-4 mb-5"
  id="leaderboard-button-tabs"
>
  <button
    className="p-3 w-full bg-purple_medium rounded-md"
    onClick={[Function]}
    type="button"
  >
    <p
      className="text-base font-medium text-white"
    >
      Weekly
    </p>
  </button>
  <button
    className="p-3 w-full hover:bg-medium_grey hover:rounded-md"
    onClick={[Function]}
    type="button"
  >
    <p
      className="text-base font-medium text-white"
    >
      All time
    </p>
  </button>
</div>
`;
